### ISSUE TITLE
Replace deprecated string functions with str methods in t.w.p.irc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5329]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Wolf reported|
|-|-|
|Trac ID|trac#5329|
|Type|enhancement|
|Created|2011-10-22 06:58:07Z|
|Branch|https://github.com/twisted/twisted/tree/replace-irc-string-5329-2|

The attached patch replaces deprecated function from the string module with the corresponding str methods in t.w.p.irc.

Attachments:

* [deprecated_string_functions.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4fd/4fd02aaf539880819f1a582d31787faea7dfa21d/4baf29fe140e0dbeadd7092a7ee25d5aaed0fcea.diff) (10140 bytes) - added by Wolf on 2011-10-22 06:58:31Z - 
* [string-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4fd/4fd02aaf539880819f1a582d31787faea7dfa21d/c6dfd32063a1dcc4cdf400b3275a962f8f3f5104.diff) (16851 bytes) - added by Wolf on 2011-11-05 01:29:20Z - 
* [issue5329.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4fd/4fd02aaf539880819f1a582d31787faea7dfa21d/abc189aac01a650fdf9e0f717c33f69ec99ad195.diff) (18981 bytes) - added by Wolf on 2013-03-26 18:04:35Z - 
* [issue5329-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4fd/4fd02aaf539880819f1a582d31787faea7dfa21d/c33a94171a12fa30d13949237b8bd1532af30b0a.diff) (20309 bytes) - added by Wolf on 2013-04-22 04:08:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5329 5329
type__enhancement enhancement
reporter__Wolf Wolf
priority__low low
milestone__ 
branch__branches_replace_irc_string_5329_2 branches/replace-irc-string-5329-2
branch_author__thijs__Wolf thijs, Wolf
status__closed closed
resolution__fixed fixed
component__words words
keywords__irc irc
time__1319266687000000 1319266687000000
changetime__1372515747000000 1372515747000000
version__None None
owner__thijs thijs
cc__ralphm cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to Wolf|
|-|-|

Thanks for your work on this.  A lot of the changed code is untested.  Do you think you could add some unit tests to exercise it?

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|Wolf set owner to @exarkun|
|-|-|

Attached patch adds tests for most of the changes.  A few changes are still untested.
The patch includes a test client that collects in a list the names of the methods and the arguments passed, and two classes to test the irc_* and ctcpQuery_* methods.
All the tests pass with and without the changes to t.w.p.irc.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to Wolf|
|-|-|

Thanks.  The tests need to conform to the testing standard, ie they need docstrings, they need to be formatted appropriately, etc.

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|Wolf removed owner|
|-|-|

I now added the docstrings to the tests.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince set owner to Wolf|
|-|-|

Thanks for your contribution, particularly the test. A couple of minor points:
 1. Untested changes:
   - sendMessage
   - dccDescribe
      - Also, python 1.5 is no longer supported, so I think the 'map(int, ...)' can go as well.
 2. `badMessage` should probably `log.err` with a `twisted.python.Failure` constructed from its arguments.
 3. (minor) The class docstrings for tests should have `"""` on separate lines. And there should be two blank lines between methods.
 4. The tests use `self.patch` for modifying an instance. There is no reason to do that, since it will be destroyed at the the end of the test anyway.

Please resubmit for review after addressing the above points.

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|Wolf removed owner|
|-|-|

I addressed the comments and attached a new patch.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [38286]) Branching to 'replace-irc-string-5329'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [38287]) apply issue5329-2.diff, add news file. refs [#5329](https://github.com/twisted/twisted/issues/5329)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to Wolf|
|-|-|

Thanks for your patch Wolf and for keeping this ticket active for 18 months now, looks like we're almost there :)

In `twisted/words/test/test_irc.py`:
 1. `test_sendMessage`:
   1. Thanks for testing `sendMessage`. could you also add a simple test for the 2 cases that a `ValueError` is raised? Should be very straightforward.
   1. The docstring should be more descriptive, what kind of sanity check is this?
 1. `CollectorClient`:
   1. `methods_list` should be documented using `@param` and `@type` instead of documenting it in an inline comment.
 1. `TestCTCPQuery`:
   1. Many of the tests have a docstring that starts with 'Test that', but you should avoid this (because we already know its a test/testing) and describe only the behavior or outcome of the test.
 1. `dccDescribeTests`:
   1. This class name should start with a capital
   1. Same comment about the docstring of the test as in comment 3.1.

In `twisted/words/protocols/irc.py`:
 1. line 1772 is untested
 1. line 2396 is untested. You also change the error message here, why is that? This kind of new feature would have to be addressed in a separate ticket.
 1. line 2662 is untested
 1. the comment above line 2726 can go

I put your patch in a branch and forced a new [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/replace-irc-string-5329). Please supply future patches against that branch, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#note_12):
> In `twisted/words/protocols/irc.py`:
>  1. line 1772 is untested
>  1. line 2396 is untested. You also change the error message here, why is that? This kind of new feature would have to be addressed in a separate ticket.

I now see that tom prince suggested this change earlier, so ignore my comment except for the fact that this line should be tested as well, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

(In [38938]) Branching to 'replace-irc-string-5329-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [38941]) address review comments, refs [#5329](https://github.com/twisted/twisted/issues/5329)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra removed owner|
|-|-|

Merged the branch forward and addressed most review comments, see below. Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/replace-irc-string-5329-2) and fixed the twistedchecker errors.

Replying to [tom.prince](#note_8):
>  2. `badMessage` should probably `log.err` with a `twisted.python.Failure` constructed from its arguments.

Can this be a new ticket?


Replying to [thijs](#note_12):
> In `twisted/words/test/test_irc.py`:
>  1. `test_sendMessage`:
>    1. Thanks for testing `sendMessage`. could you also add a simple test for the 2 cases that a `ValueError` is raised? Should be very straightforward.

Added two additional tests.

>    1. The docstring should be more descriptive, what kind of sanity check is this?

Fixed.

>  1. `CollectorClient`:
>    1. `methods_list` should be documented using `@param` and `@type` instead of documenting it in an inline comment.

Fixed by merging forward.

>  1. `TestCTCPQuery`:
>    1. Many of the tests have a docstring that starts with 'Test that', but you should avoid this (because we already know its a test/testing) and describe only the behavior or outcome of the test.

Fixed.

>  1. `dccDescribeTests`:
>    1. This class name should start with a capital

Fixed.

> In `twisted/words/protocols/irc.py`:
>  1. line 2396 is untested. You also change the error message here, why is that? This kind of new feature would have to be addressed in a separate ticket.

Reverted the change.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince set owner to @thijstriemstra|
|-|-|

Thanks for working on this

 1. `test_sendMessage`: Words like "correct" and "valid" should be avoided in test docstrings, the docstring should say what it means for the query string to be valid (see [#6301](https://github.com/twisted/twisted/issues/6301))
   - The same is true for `TestCTCPQuery`
>>  2. `badMessage` should probably `log.err` with a `twisted.python.Failure` constructed from its arguments.  
>Can this be a new ticket?
Sure.
- Since it will be addressed in a seperate ticket, we should consider deprecating `badMessage`, and replacing it with a function that takes a `Failure` instead.

Please merge after addressing 1, and filing a ticket for 2.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra commented|
|-|-|

(In [38978]) address review comments, refs [#5329](https://github.com/twisted/twisted/issues/5329)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_19"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review.

Replying to [tom.prince](#note_17):
> Please merge after addressing 1, and filing a ticket for 2.

Opened [#6605](https://github.com/twisted/twisted/issues/6605).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [38979]) Merge replace-irc-string-5329-2: Replaced usage of deprecated string functions in words.protocols.irc.

Author: Wolf, thijs
Reviewer: exarkun, tom.prince
Fixes: [#5329](https://github.com/twisted/twisted/issues/5329)

**Comment by twisted-trac** [comment]
|<img alt="Wolf's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|Wolf commented|
|-|-|

Thanks!
